File: framework/languages/java/samples/java/TicTacToe/Server/src/main/java/systems/zlink/samples/tictactoe/server/play/infrastructure/zlink/spots/entryspot/handlers/PlayActorJoinGameHandler.java
package systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkSpotActorSend;
import systems.zlink.samples.tictactoe.server.configuration.SampleNames;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.actors.PlayActor;
import systems.zlink.samples.tictactoe.server.play.infrastructure.zlink.spots.entryspot.PlayEntrySpot;
import systems.zlink.samples.tictactoe.shared.contracts.JoinGameReq;
import systems.zlink.samples.tictactoe.shared.contracts.TicTacToeGameJoinReq;

@ZLinkHandlerGroup(SampleNames.PlayActor)
// --8<-- [start:doc-join-defer]
public final class PlayActorJoinGameHandler {
    @ZLinkSpotActorSend
    public java.util.concurrent.CompletionStage<Void> joinGame(
        PlayEntrySpot entrySpot,
        PlayActor actor,
        ZLinkMessageContext context,
        JoinGameReq request) {
        actor.trackDeferredJoin(request.roomId());
        actor.context()
            .joinSpot(request.roomId(),
                new TicTacToeGameJoinReq(request.roomId(), actor.requirePlayer()))
            .timeout(SampleNames.RequestTimeout)
            .defer();
        return java.util.concurrent.CompletableFuture.completedFuture(null);
    }
}
// --8<-- [end:doc-join-defer]

File: framework/languages/node/samples/TicTacToe.Ts/Server/Play/Infrastructure/ZLink/Spots/EntrySpot/Handlers/play-actor-join-game-handler.ts
import { zlinkEntrySpotActorRequestHandler } from '@zlink-systems/nestjs';
import type {
  ZLinkEntrySpotActorRequestHandler,
  ZLinkMessageContext
} from '@zlink-systems/framework';
import { PlayActor } from '../../../Actors/play-actor';
import { PlayEntrySpot } from '../play-entry-spot';
import type {
  JoinGameReq,
  JoinGameRes,
  TicTacToeGameJoinReq
} from '../../../../../../../Shared/Contracts/messages';
import { GameStatus, joinGameRes } from '../../../../../../../Shared/Contracts/messages';

@zlinkEntrySpotActorRequestHandler({
  actor: () => PlayActor,
  entrySpot: () => PlayEntrySpot,
  packetName: 'JoinGameReq'
})
// --8<-- [start:doc-join-defer]
class PlayActorJoinGameHandler
  implements ZLinkEntrySpotActorRequestHandler<PlayEntrySpot, PlayActor, JoinGameReq, JoinGameRes> {
  async handle(
    _spot: PlayEntrySpot,
    actor: PlayActor,
    context: ZLinkMessageContext,
    request: JoinGameReq
  ): Promise<JoinGameRes> {
    void context;
    const joinRequest: TicTacToeGameJoinReq = {
      roomId: request.roomId,
      player: {
        actorId: actor.actorId,
        displayName: actor.displayName,
        level: actor.level,
        wins: actor.wins
      }
    };
    actor.context
      .joinSpot(request.roomId, joinRequest)
      .defer();
    actor.roomId = request.roomId;
    return joinGameRes({
      roomId: request.roomId,
      board: '.........',
      status: GameStatus.WaitingForPlayers,
      winner: null,
      nextTurn: '',
      xActorId: null,
      oActorId: null,
      lastMoveActorId: null,
      lastMoveCell: null
    });
  }
}
// --8<-- [end:doc-join-defer]

export { PlayActorJoinGameHandler };
